File: clients/protocol-client/src/types/generation.ts
import { CellComponentSelector } from "./cell";

type GenerationStepKind =
  | "RandGen"
  | "InitializeCell"
  | "PairwiseStep"
  | "PairwiseMerge"
  | "Finalize";

const GenerationStepKind = {
  RAND_GEN: "RandGen" as const,
  INITIALIZE_CELL: "InitializeCell" as const,
  PAIRWISE_STEP: "PairwiseStep" as const,
  PAIRWISE_MERGE: "PairwiseMerge" as const,
  FINALIZE: "Finalize" as const,
};

type GenerationPhase =
  | "NewlyCreated"
  | "PreInitialize"
  | "CellInitialized"
  | "PreMerge"
  | "Finalized";

type GenerationCellDatumId =
  | { RandGen: {} }
  | { Selector: CellComponentSelector };

const GenerationCellDatumId = {
  toStringKey(id: GenerationCellDatumId): string {
    if ("RandGen" in id) {
      return "RandGen";
    } else {
      const { word, component } = id.Selector;
      return `Selector:${word}:${component}`;
    }
  },

  equal(a: GenerationCellDatumId, b: GenerationCellDatumId): boolean {
    return (
      GenerationCellDatumId.toStringKey(a)
        === GenerationCellDatumId.toStringKey(b)
    );
  }
};

export { GenerationStepKind, GenerationPhase, GenerationCellDatumId };

File: src/world/generation.rs
use super::cell_data::CellComponentSelector;


#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) enum GenerationStepKind {
  RandGen,         // NewlyCreated -> PreInitialize
  InitializeCell,  // PreInitialize -> CellInitialized
  PairwiseStep,    // CellInitialized -> PreMerge
  PairwiseMerge,   // PreMerge -> CellInitialized
  Finalize,        // CellInitialized -> Finalized
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) enum GenerationPhase {
  NewlyCreated,
  PreInitialize,
  CellInitialized,
  PreMerge,
  Finalized
}
impl GenerationPhase {
  pub(crate) fn to_str(&self) -> &'static str {
    match self {
      GenerationPhase::NewlyCreated => "NewlyCreated",
      GenerationPhase::PreInitialize => "PreInitialize",
      GenerationPhase::CellInitialized => "CellInitialized",
      GenerationPhase::PreMerge => "PreMerge",
      GenerationPhase::Finalized => "Finalized",
    }
  }
}

#[derive(Clone, Debug)]
#[derive(serde::Serialize, serde::Deserialize)]
pub(crate) enum GenerationCellDatumId {
  /** Value of named randgen field. */
  RandGen {},

  /** Value of map cell format word component */
  Selector(CellComponentSelector),
}
